client: support no_pivot_root in exec driver configuration

th0m opened the pull request:
Adds a new `no_pivot_root` plugin option to the `exec` plugin.
It defaults to `false`, changing it to `true` will pass the `NoPivotRoot` configuration option to `libcontainer` which will fall back to using the `msMoveRoot` function for isolation.
This is useful for systems where the root is on a ramdisk.

Fixes #7136

hashicorp-cla commented:
[![CLA assistant check](https://cla.hashicorp.com/pull/badge/not_signed)](https://cla.hashicorp.com/hashicorp/nomad?pullRequest=7149)


Thank you for your submission! We require that all contributors sign our Contributor License Agreement ("CLA") before we can accept the contribution. [Read and sign the agreement](https://cla.hashicorp.com/hashicorp/nomad?pullRequest=7149)

[Learn more about why HashiCorp requires a CLA and what the CLA includes](https://www.hashicorp.com/cla)

<sub>Have you signed the CLA already but the status is still pending? [Recheck it](https://cla.hashicorp.com/check/hashicorp/nomad?pullRequest=7149).</sub>

notnoop commented on drivers/exec/driver.go:
```diff
@@ -88,6 +93,9 @@ type Driver struct {
 	eventer *eventer.Eventer
 
+	// config is the driver configuration set by the SetConfig RPC
+	config *Config
```
`d.config` might be nil in tests - hence one test failure.  You may consider avoid using the pointer here.

```suggestion
	config Config
```

notnoop commented on drivers/shared/executor/proto/executor.proto:
```diff
@@ -28,9 +28,10 @@ message LaunchRequest {
-    repeated hashicorp.nomad.plugins.drivers.proto.Mount mounts = 11;
-    repeated hashicorp.nomad.plugins.drivers.proto.Device devices = 12;
-    hashicorp.nomad.plugins.drivers.proto.NetworkIsolationSpec network_isolation = 13;
+    bool no_pivot_root = 11;
```
I would insert append this to the end as item 14.  For backward compatibility, field numbers don't change once they are assigned.

th0m commented:
Thanks for the review @notnoop. I'll add the integration test and address your comments.
This PR relates to issue #7136

th0m commented:
I added a test and addressed your comments @notnoop, let me know if that looks good and thanks again for the feedback!
I still have a failed build for {{lint-go}} which I think was due to the fact I am using protobuf 3.11.3, I downgraded to 3.6.1 which is what CI seems to be using and the {{make proto}} output lined up with what was in CI now.

th0m commented:
All checks are green, looks like the `protobuf` downgrade before generating proto files fixed my `lint-go` build problem.

notnoop commented on website/pages/docs/drivers/exec.mdx:
```diff
@@ -93,6 +93,13 @@ If you are receiving the error:
+* `no_pivot_root` - Defaults to `false`. Changing this to `true` will pass the
+  `NoPivotRoot` configuration option to `libcontainer` which will fall back to
+  using the `msMoveRoot` function for isolation. This is useful for systems
+  where the root is on a ramdisk.
```
Some of this feels like an implementation details (e.g. the bits about libcontainer).  How do you feel about something like:

```md
* `no_pivot_root` - Defaults to `false`. When `true`, the driver uses `chroot` for file system
  isolation without `pivot_root`.  This is useful for systems where the root is on a ramdisk.
```

th0m replied:
Sure, that sounds good! I just updated the PR.
